src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
Ah, I see. So the try-catch block is there to handle `.get()`'s exceptions.

Then, the next question that arises is, why is `loadRepositoryFromStore` handling the task of downloading data from source? The name `loadRepositoryFromStore` is unclear as it doesn't imply that `loadRepositoryFromStore` will trigger the download.

So, you can either change the name to `loadRepositoryFromStore` to something clearer and longer (which is probably not desirable), or maybe you can do error handling at a higher level in the call chain, like in `openRepository`?